community/kernel/src/main/java/org/neo4j/kernel/impl/store/GeometryType.java

```diff
@@ -0,0 +1,164 @@
+                {
+                    int dimension = getDimension( firstBlock );
+                    if ( dimension > 3 )
+                    {
+                        return PropertyType.BLOCKS_USED_FOR_BAD_TYPE_OR_ENCODING;
+                    }
+                    return 1 + dimension;
+                }
+            };
+
+    public final int gtype;
+
+    GeometryType( int gtype )
+    {
+        this.gtype = gtype;
+    }
+
+    public abstract Value decode( CoordinateReferenceSystem crs, int dimension, long[] valueBlocks, int offset );
+
+    // TODO this might not work this way for Polygons or other geometries that do not
+    // know their length just from the firstBlock. Since we will probably use dynamic property
+    // stores, it might be that the answer here is just 1.
+    public abstract int calculateNumberOfBlocksUsedForGeometry( long firstBlock );
+
+    private static int getGeometryType( long firstBlock )
+    {
+        return (int) ((firstBlock & 0x00000000F0000000L) >> 28);
+    }
+
+    private static int getDimension( long firstBlock )
+    {
+        return (int) ((firstBlock & 0x0000000F00000000L) >> 32);
+    }
+
+    private static int getCRSTable( long firstBlock )
+    {
+        return (int) ((firstBlock & 0x000000F000000000L) >> 36);
+    }
+
+    private static int getCRSCode( long firstBlock )
+    {
+        return (int) ((firstBlock & 0x00FFFF0000000000L) >> 40);
+    }
+
+    private static boolean isFloatPrecision( long firstBlock )
+    {
+        return ((firstBlock & 0x0100000000000000L) >> 56) == 1;
+    }
+
+    public static int calculateNumberOfBlocksUsed( long firstBlock )
+    {
+        GeometryType geometryType = find( getGeometryType( firstBlock ) );
+        if ( geometryType == null )
+        {
+            return PropertyType.BLOCKS_USED_FOR_BAD_TYPE_OR_ENCODING;
+        }
+        else
+        {
+            return geometryType.calculateNumberOfBlocksUsedForGeometry( firstBlock );
+        }
+    }
+
+    private static GeometryType find( int gtype )
+    {
+        for ( GeometryType type : GeometryType.values() )
+        {
+            if ( type.gtype == gtype )
+            {
+                return type;
+            }
+        }
+        return null;
+    }
+
+    public static Value decode( PropertyBlock block )
+    {
+        return decode( block.getValueBlocks(), 0 );
+    }
+
+    public static Value decode( long[] valueBlocks, int offset )
+    {
+        long firstBlock = valueBlocks[offset];
+        int gtype = getGeometryType( firstBlock );
+        int dimension = getDimension( firstBlock );
+
+        if ( isFloatPrecision( firstBlock ) )
+        {
+            throw new UnsupportedOperationException( "Float precision is unsupported in Geometry properties" );
+        }
+        if ( dimension > 3 )
+        {
+            throw new UnsupportedOperationException( "Points with more than 3 dimensions are not supported in the PropertyStore" );
```

Comment:
i would just say that "Points with more than 3 dimensions are not supported" since most of the supported/unsupported things are property format related and those can be quite different independently from type of store.